### ISSUE TITLE
Clone method implemented incorrectly for TypeInfoFactoryDataType

### ISSUE THREAD

**Issue by ghidra1** [astrelsky/Ghidra-Cpp-Class-Analyzer#65]
The clone method should always return an implementation with the supplied DataTypeManager.  This is not the case.  I suggest the following implementation:
```
	public DataType clone(DataTypeManager dtm) {
		if (dtm == getDataTypeManager()) {
			return this;
		}
		return new TypeInfoFactoryDataType(dtm);
	}
```
As it currently stands if the dtm is null `this` will be returned.  Not sure if you have other datatypes with a similar issue.  A user accidently resolved the type into a program which produced various issues for the DataTypeManager and the tree.

You may be able to reproduce the issue by dragging your datatype from the BuiltInManager to a program.  Save the program and re-open then try to find it in the program's datatype tree.

**Comment by astrelsky** [comment]
I saw that issue this morning, saw the "TypeInfoDataType" and did not put 2 and 2 together. While it was a few years ago now, the way it is implemented was copied from how it was in other builtin types at the time. Unless that has changed, the issue might also be present in some of Ghidra's builtin types and may need to be checked.

I do recall doing something shady at one point to prevent a type from being moved to anywhere other than the BuiltinDataTypeManager but I'm not sure if the class containing it still exists. I'll check tonight and remove it if it does.

**Comment by astrelsky** [comment]
Now I'm the one lying to you. Disregard the deleted message.

The way it is implemented was supposed to make it impossible to be moved outside of the `BuiltinDataTypeManager`. If it is moved to a file archive and then used where the plugin is not installed,  then what happens?

Either way I'm going to remove this type. This behavior should have been implemented as a `DockingAction`.

**Comment by ghidra1** [comment]
Any `BuiltIn` available to `BuiltInDataTypeManager` can be resolved into a program.  No way to prevent that.  It is possible to have a protected `DataType` (not `BuiltIn`) although this should generally be limited to types used by another Dyanmic datatype which is BuiltIn and public (e.g., `AndroidElfRelocationTableDataType` uses various protected types).   How this is done depends on how the DataType will be used.  Hiding techniques include:
- Do not use `DataType` class name suffix
- Non-public visibility

**Comment by astrelsky** [comment]
"Fixed" by 187492cd49e1b50c8102bf929138159e9d5085f4
